src/WebJobs.Extensions.DurableTask/Listener/TaskHttpActivityShim.cs

```diff
@@ -0,0 +1,112 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading.Tasks;
+using DurableTask.Core;
+using Microsoft.Extensions.Primitives;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask.Listener
+{
+    internal class TaskHttpActivityShim : TaskActivity
+    {
+        private readonly HttpClient httpClient;
+        private readonly DurableTaskExtension config;
+
+        public TaskHttpActivityShim(
+            DurableTaskExtension config,
+            HttpClient httpClientFactory)
+        {
+            this.config = config ?? throw new ArgumentNullException(nameof(config));
+            this.httpClient = httpClientFactory;
+        }
+
+        public override string Run(TaskContext context, string input)
+        {
+            // This won't get called as long as we've implemented RunAsync.
+            throw new NotImplementedException();
+        }
+
+        public async override Task<string> RunAsync(TaskContext context, string rawInput)
+        {
+            HttpRequestMessage requestMessage = await CreateHttpRequestMessageAsync(rawInput);
+            HttpResponseMessage response = await this.httpClient.SendAsync(requestMessage);
+            DurableHttpResponse durableHttpResponse = await CreateDurableHttpResponseAsync(response);
+
+            string serializedOutput = MessagePayloadDataConverter.HttpConverter.Serialize(durableHttpResponse, true);
+            return serializedOutput;
+        }
+
+        private static async Task<DurableHttpResponse> CreateDurableHttpResponseAsync(HttpResponseMessage response)
+        {
+            DurableHttpResponse durableHttpResponse = new DurableHttpResponse(response.StatusCode);
+            durableHttpResponse.Headers = CreateStringValuesHeaderDictionary(response.Headers);
+            durableHttpResponse.Content = await response.Content.ReadAsStringAsync();
+            return durableHttpResponse;
+        }
+
+        private static IDictionary<string, StringValues> CreateStringValuesHeaderDictionary(IEnumerable<KeyValuePair<string, IEnumerable<string>>> headers)
+        {
+            IDictionary<string, StringValues> newHeaders = new Dictionary<string, StringValues>();
+            if (headers != null)
+            {
+                foreach (var header in headers)
+                {
+                    foreach (var headerValue in header.Value)
```

Comment:
I think you can entirely remove this inner loop. `StringValues` has a constructor that takes a `string[]`. I think then you can replace this inner loop with `newHeaders[header.Key] = new StringValues(header.Value.ToArray());`.